### PR TITLE
FEATURE: Add server timing header

### PR DESCRIPTION
**Pull request by Sebobo**
This header will create additional output in the browser dev tools.
Each timing will be added as entry in the timings tab and helps devs
to understand which part of the app took how long.

Initially it will provide the duration of the request process, fusion rendering time and sql execution time.

But via the DebugService additional timings can be added by the dev.

Example:

server-timing-example

### CODE REVIEW

File: Classes/Service/DebugService.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * @Flow\Scope("singleton")
+ */
+class DebugService {
+    protected $startRenderAt;
+    protected $endRenderAt;
+    protected $sqlLoggingStack;
+    protected $customTimings;
+
+    public function start(): void {
```
Comment (johannessteu) [inline Classes/Service/DebugService.php:17]: Maybe rename to a more declarative function name? `startRenderCount` or something like this? This Service could be used for more cases as well and `start` might me miss leading later on

File: Classes/Service/DebugService.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        $this->startRenderAt = microtime(true) * 1000;
+    }
+
+    public function end(): float {
```
Comment (johannessteu) [inline Classes/Service/DebugService.php:21]: Just like `start()`

File: Classes/Service/DebugService.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        return $this->endRenderAt = microtime(true) * 1000;
+    }
+
+    public function setCustomTiming(string $name, float $value): void {
```
Comment (johannessteu) [inline Classes/Service/DebugService.php:25]: Rename to addCustomTiming?

**Comment by johannessteu** [comment]
I like it! beside some lint issues and naming discussion i thing this might be good to go once some documentation is adde @Sebobo !
Thanks for this!

**Comment by Sebobo** [comment]
Thx!

Yes this was just the first draft and I wanted the feedback whether the direction makes sense.
I will change the naming.

I also want to add the actual application time incl. bootstrap and cache generation. Currently with a cold cache the request might take some seconds but the timings still show < 1s. This should be possible via `$_SERVER['REQUEST_TIME_FLOAT']`.
Or do we have something in the Framework for that which I missed?

Also I thought about having a separate setting for enabling this header. Because I would like to always have this header but not necessarily the output in the html. What do you think?

**Comment by Sebobo** [comment]
There are several more possibilities in the specification: https://w3c.github.io/server-timing/#examples

* [ ] So adding a description is possible and should be added.
* [ ] Some characters should not be contained in the key and description to prevent breaking the format.
* [ ] Metrics without value are possible (only the key  is shown e.g. "Cache miss").
* [ ] Metrics without value can have a description.

They mention in the specification that those metrics could be used for tracking and monitoring purposes. So usually you don't want internal timing data visible to the user for security purposes. But lately I think about implementing server side tracking for the Matomo package. And there this information could be appended and worked with in the analysis.

**Comment by Sebobo** [comment]
@johannessteu ok I thinks everything regarding the specification is in there now. I added more configuration options and changed the naming.

**Comment by johannessteu** [comment]
I'll have a look into the ci issue and merge it afterwards :)

**Comment by Sebobo** [comment]
Did you have a change to check the CI issue?

**Comment by johannessteu** [comment]
Moved over to #93

**Pull request closed without merging**